File: backend/enum-jpa-validation-app/src/test/java/com/example/enumapp/domain/order/OrderFlexibleSearchCriteriaBuilderTest.java
package com.example.enumapp.domain.order;

import com.example.enumapp.web.dto.OrderFlexibleSearchRequest;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;

import static org.assertj.core.api.Assertions.assertThat;

class OrderFlexibleSearchCriteriaBuilderTest {

    @Test
    void emptyRequest_buildsEmptyCriteria() {
        OrderFlexibleSearchCriteria criteria = OrderFlexibleSearchCriteriaBuilder.from(new OrderFlexibleSearchRequest());
        assertThat(criteria.isEmpty()).isTrue();
    }

    @Test
    void blankStringsAndNulls_areIgnored() {
        OrderFlexibleSearchRequest request = new OrderFlexibleSearchRequest();
        request.setCustomerName("  ");
        request.setOrderDate("");
        request.setFromDate(null);
        request.setMinQuantity(null);
        assertThat(OrderFlexibleSearchCriteriaBuilder.from(request).isEmpty()).isTrue();
    }

    @Test
    void orderDate_winsOverBetweenFields() {
        OrderFlexibleSearchRequest request = new OrderFlexibleSearchRequest();
        request.setOrderDate("2026-08-06");
        request.setFromDate("2026-01-01");
        request.setToDate("2026-12-31");

        OrderFlexibleSearchCriteria criteria = OrderFlexibleSearchCriteriaBuilder.from(request);
        assertThat(criteria.getCreatedFrom()).isEqualTo(LocalDateTime.of(2026, 8, 6, 0, 0));
        assertThat(criteria.getCreatedTo()).isEqualTo(LocalDateTime.of(2026, 8, 7, 0, 0));
        assertThat(criteria.isCreatedToInclusive()).isFalse();
    }

    @Test
    void mapsEnumNumberAndString() {
        OrderFlexibleSearchRequest request = new OrderFlexibleSearchRequest();
        request.setCustomerName(" Kim ");
        request.setStatus(OrderStatus.PENDING);
        request.setPayMethod(PayMethod.CARD);
        request.setMinQuantity(3);

        OrderFlexibleSearchCriteria criteria = OrderFlexibleSearchCriteriaBuilder.from(request);
        assertThat(criteria.getCustomerName()).isEqualTo("Kim");
        assertThat(criteria.getStatus()).isEqualTo(OrderStatus.PENDING);
        assertThat(criteria.getPayMethod()).isEqualTo(PayMethod.CARD);
        assertThat(criteria.getMinQuantity()).isEqualTo(3);
    }
}

File: backend/enum-jpa-querydsl-kotlin-app/src/test/kotlin/com/example/enumapp/domain/order/OrderFlexibleSearchCriteriaBuilderTest.kt
package com.example.enumapp.domain.order

import com.example.enumapp.web.dto.OrderFlexibleSearchRequest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

class OrderFlexibleSearchCriteriaBuilderTest {

    @Test
    fun emptyRequest_buildsEmptyCriteria() {
        val criteria = OrderFlexibleSearchCriteriaBuilder.from(OrderFlexibleSearchRequest())
        assertThat(criteria.isEmpty()).isTrue()
    }

    @Test
    fun blankStringsAndNulls_areIgnored() {
        val request = OrderFlexibleSearchRequest().apply {
            customerName = "  "
            orderDate = ""
            fromDate = null
            minQuantity = null
        }
        assertThat(OrderFlexibleSearchCriteriaBuilder.from(request).isEmpty()).isTrue()
    }

    @Test
    fun orderDate_winsOverBetweenFields() {
        val request = OrderFlexibleSearchRequest().apply {
            orderDate = "2026-08-06"
            fromDate = "2026-01-01"
            toDate = "2026-12-31"
        }

        val criteria = OrderFlexibleSearchCriteriaBuilder.from(request)
        assertThat(criteria.createdFrom).isEqualTo(LocalDateTime.of(2026, 8, 6, 0, 0))
        assertThat(criteria.createdTo).isEqualTo(LocalDateTime.of(2026, 8, 7, 0, 0))
        assertThat(criteria.createdToInclusive).isFalse()
    }

    @Test
    fun mapsEnumNumberAndString() {
        val request = OrderFlexibleSearchRequest().apply {
            customerName = " Kim "
            status = OrderStatus.PENDING
            payMethod = PayMethod.CARD
            minQuantity = 3
        }

        val criteria = OrderFlexibleSearchCriteriaBuilder.from(request)
        assertThat(criteria.customerName).isEqualTo("Kim")
        assertThat(criteria.status).isEqualTo(OrderStatus.PENDING)
        assertThat(criteria.payMethod).isEqualTo(PayMethod.CARD)
        assertThat(criteria.minQuantity).isEqualTo(3)
    }
}
